src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -247,15 +247,24 @@ private Task WriteFormattedInt32Async(int value)
                 await WriteStringAsync(request.Method.Method).ConfigureAwait(false);
                 await WriteByteAsync((byte)' ').ConfigureAwait(false);
 
-                if (_usingProxy)
+                if (request.IsConnectMethod)
                 {
-                    // Proxied requests contain full URL
-                    Debug.Assert(request.RequestUri.Scheme == Uri.UriSchemeHttp);
-                    await WriteBytesAsync(s_httpSchemeAndDelimiter).ConfigureAwait(false);
-                    await WriteAsciiStringAsync(request.RequestUri.IdnHost).ConfigureAwait(false);
+                    // RFC 7231 #section-4.3.6.
+                    // Write only CONNECT foo.com:345 HTTP/1.1
+                    Debug.Assert(request.HasHeaders && request.Headers.Host != null);
```

Comment:
The user could construct this request themselves, so rather than Assert we should throw here.

Reply:
btw it would be easy to fabricate it as well - we do that in other cases.

Reply:
How would we fabricate it?

Reply:
What authority?  The request Uri?  In this case, that will contain the proxy Uri.

Reply:
Yes, authority from request URI. That is what I see when existing handlers wants to talk to proxy. Host header and request URI authority are identical. If somebody for example wants to talk via proxy without using IWebProxy interface, that is exactly what they would do. 
Anybody can of course set the header to what ever they want so the question really is what should be default action if they don't.